### ISSUE TITLE
Minimal Baskets/Regen Ledger 3.0

### ISSUE THREAD

**Issue by aaronc** [regen-network/regen-ledger#732]
Guidelines:

- **one PR with unit tests per checklist item!**
- **all PRs against `release/v3.0.x`** unless otherwise noted
- **please put your name next to the checkbox item or issue you're tackling** and _also_ **communicate on discord**
- let's keep these small with isolated unit tests so there's a quick turn-around (< 1 day) on each item
- feel free to open multiple PRs for separate items, but review other people's PRs before you start a new one
- to keep momentum, we pass off PRs that we don't have the bandwidth to finish quickly

## BLOCKERS
  - [x] does `MsgCreate` ValidateBasic needs fee validation? (only check for valid Coins.Validate() which accepts empty coins)
  - [x] add validation & construction for credit class name & basket denom
  
## Todo
- [x] create `release/v3.0.x` branch
- [x] basic proto files #731 (against master)
- [x] backport just gogo proto codegen (not api module or .proto files) from #731 to `release/v3.0.x` @robert-zaremba 
- [x] tag `api` module on master after #731 is merged: `api/v0.1.0` tagged
- [x] #735 #734 @aaronc 
- [x] events.proto on master @ryanchristo https://github.com/regen-network/regen-ledger/pull/736
- [x] events.proto release backport (without api module) #743
- [x] query.proto on master + backport (only gogo proto codegen) @technicallyty https://github.com/regen-network/regen-ledger/pull/738
- [x] query.proto release backport (without api module) #743
- [x] #740 @robert-zaremba 
- [x] add `BasketCreationFee` param support (is currently breaking tests on `release/v3.0x`) @aaronc #741 
- [x] add rolling date window (based on conversations with @clevinson) @aaronc #744 
  - [x] tag api package: @aaronc `api/v0.2.0` tagged
  - [x] backport #744 to release/3.0 @robert-zaremba 
  - [x] update `ValidateBasic()` on `MsgCreate` to support rolling window -- validate that Duration >= 1 day @robert-zaremba 
  - [x] ensure MsgServer implementation supports rolling date window.
- `sdk.Msg` + `sdk.LegacyMsg` implementations
  - [x] `MsgCreate` @robert-zaremba #745 
  - [x] `MsgPut` @technicallyty https://github.com/regen-network/regen-ledger/pull/747
  - [x] `MsgTake` @ryanchristo https://github.com/regen-network/regen-ledger/pull/748
- [x] codec registration including amino (with names shorter than 40 characters) @aaronc #765
- `MsgServer` (each of these done with mock-based unit tests initially)
  - [x] `Create` @robert-zaremba #762
  - [x] `Put` @technicallyty #758 
  - [x] `Take` @aaronc #746 
- [x] genesis (should be very simple and use orm `ImportJSON`/`ExportJSON`) @aaronc #764
- queries
  - [x] `Basket` @ryanchristo #753
  - [x] `Baskets` @aaronc #761 
  - [x] `BasketBalance` @aaronc  #751 
  - [x] `BasketBalances` @technicallyty https://github.com/regen-network/regen-ledger/pull/763
-  CLI (full integration tests not necessarily needed)
   - tx
     - [x] `create-basket` @ryanchristo #754
     - [x] `add-to-basket` @aleem1314 #750
     - [x] `take-from-basket` @aleem1314 #749
   - query
     - [x]  `basket` @technicallyty #766 
     - [x] `baskets` @technicallyty #767 
     - [x] `basket-balance` @aleem1314 #760 
     - [x] `basket-balances` @aleem1314 #760 
 
### Beta Blocking
- [x] Finalize MsgCreate name, display name and denom validation once we confirm the requirements #772 @robert-zaremba  
- [x] Upgrade handler (no migrations needed) https://github.com/regen-network/regen-ledger/pull/775
  - [x] Confirm default create basket fee - 1000REGEN?
- [x] update proto to remove `MsgCreate.prefix` and add `Basket.name` unique index @aaronc #774 
- [x] v2.2 update and new api release after #774 @robert-zaremba 
- [x] register basket `SubModuleName` as a module account properly (app.go, etc.) #771 
- [x] Set the CoinDenomRegxp (prob also in app.go?) #771 

### RC Blocking
- [x] Confirm proper event usage (EventRetire, EventReceive, etc.) @aaronc #778 
  - [x] Manually backport to `release/v3.0.x` and update `EventReceive` usage @robert-zaremba #794
- [x] integration tests added to https://github.com/regen-network/regen-ledger/tree/master/x/ecocredit/server/testsuite @technicallyty #771
- [x] Simulations @aleem1314 #769 
  - [x] fixing amino signing for date criteria #790 #792
  - [x] properly do `Dec` -> `Int` conversion in `Put` @aaronc #783 
- [x] Fix `stable-code-cov`: panic on GetParamSet on L34 of x/ecocredit/server/msg_server.go #784
- [x] add invariant check for ensuring bank supply of basket token equals sum of ecocredits in the basket #787
- [x] ecocredit consensus version should bump to 2 (upgrade handler) #789
- [x] register basket related endpoints #789

## QA
- [x] Testnet creation using export genesis @kaustubhkapatral 
  - export mainnet 
  - update validator set
  - migrate genesis (existing ecocredit genesis needs to go under `regen.ecocredit.v1alpha1.GenesisState` key)
- [ ] write acceptance tests @ryanchristo #800
- [x] manually testing (2+ people run this scenario from the CLI) @kaustubhkapatral, @ryanchristo, @clevinson
    - [x] Team audit and @kaustubhkapatral manual testing report: https://hackmd.io/IZittJ8fRoOf-defmaloWQ
    - [x] @ryanchristo manual testing report: https://hackmd.io/wKk_NURPRtyuqcycUdzlnw
    - [x] @clevinson manual testing report: https://hackmd.io/Novv1-fBRvy1O1TQrVVduQ
- [x] code review audit: re-review code committed to `release/3.0.x` with special emphasis on `MsgServer` and `Msg.ValidateBasic` (2 people) @aaronc 
  - [x] Locally in IDE run tests with coverage and look for coverage gaps
  - [x] do a diff btw v3.0 and v2.1 tag and read through the diff and write down any issues you see (starting with msgServer code)
---

## Nice to Have/Post Release
- [ ] #833
- [ ] #779
- [x] #834
- [x] add gas Take iteration (state machine breaking - **can't be done in point release**)  @robert-zaremba 
- [ ] #836

**Comment by aaronc** [comment]
Okay to close this? Should we move the follow-ups to separate issues? Are they getting tracked and addressed?

**Comment by ryanchristo** [comment]
> Okay to close this? Should we move the follow-ups to separate issues? Are they getting tracked and addressed?

Follow-up moved to separate issues and being tracked as part of the v4.0 upgrade. Closing.
